compiler/rustc_trait_selection/src/error_reporting/infer/need_type_info.rs

```diff
@@ -613,6 +666,130 @@ impl<'a, 'tcx> TypeErrCtxt<'a, 'tcx> {
             }),
         }
     }
+
+    fn get_suggestions(
+        &self,
+        self_ty: Ty<'tcx>,
+        def_id: DefId,
+        target_type: bool,
+        param_env: ty::ParamEnv<'tcx>,
+        args: Option<&ty::GenericArgs<'tcx>>,
+        predicate: Option<ty::Predicate<'tcx>>,
+    ) -> Vec<String> {
+        let tcx = self.infcx.tcx;
+        let mut paths = vec![];
+        let name = tcx.item_name(def_id);
+        let empty_args = |def_id| {
+            ty::GenericArgs::for_item(tcx, def_id, |param, _| {
+                // We don't want to name the arguments, we just want to give an
+                // idea of what the syntax is.
+                match param.kind {
+                    ty::GenericParamDefKind::Lifetime => tcx.lifetimes.re_erased.into(),
+                    ty::GenericParamDefKind::Type { .. } => self.next_ty_var(DUMMY_SP).into(),
+                    ty::GenericParamDefKind::Const { .. } => self.next_const_var(DUMMY_SP).into(),
+                }
+            })
+        };
+        let args = args.unwrap_or_else(|| empty_args(def_id));
+        let trait_def_id = tcx.parent(def_id);
+        tcx.for_each_relevant_impl(trait_def_id, self_ty, |impl_def_id| {
+            let impl_args = empty_args(impl_def_id);
+            let impl_trait_ref =
+                tcx.impl_trait_ref(impl_def_id).unwrap().instantiate(tcx, impl_args);
+            let impl_self_ty = impl_trait_ref.self_ty();
+            if self.infcx.can_eq(param_env, impl_self_ty, self_ty) {
+                // The expr's self type could conform to this impl's self type.
+            } else {
+                // Nope, don't bother.
+                return;
+            }
+
+            let filter = if let Some(ty::ProjectionPredicate {
+                projection_term: ty::AliasTerm { def_id, .. },
```

Comment:
How does this `DefId` differ from the one passed as an argument?

Reply:
This is the DefId of an associated type (like in the tests, `Mul::Output`, while the other `def_id` is the one for the method (in the same test, `<i16 as Into<_>::into(..)`).